File: client/common/comms.go
package common

import (
	"bytes"
	"encoding/binary"
	"errors"
	"io"
	"net"
	"time"

	log "github.com/sirupsen/logrus"
)

func sendStopedProcessingToServer(c *Client) error {
	//Send message of finished processing
	log.Infof("Sending finished processing message to server")
	msg := new(bytes.Buffer)
	err := msg.WriteByte('\f')
	if err != nil {
		return err
	}
	err = msg.WriteByte('\f')
	if err != nil {
		return err
	}
	err = sendAll(msg.Bytes(), &c.conn)
	if err != nil {
		return err
	}
	return nil
}

func sendFinishedToServer(c *Client) error {
	log.Infof("Sending finish message to server")
	msgLen := new(bytes.Buffer)
	err := msgLen.WriteByte('\n')
	if err != nil {
		return err
	}
	err = msgLen.WriteByte('\n')
	if err != nil {
		return err
	}
	err = sendAll(msgLen.Bytes(), &c.conn)
	if err != nil {
		return err
	}
	return nil
}

func requestTotalWinners(c *Client) error {
	for {
		if c.checkForOsSingal() {
			return errors.New("Received SIGTERM while waiting for total winners")
		}
		log.Infof("Sending query message to server")
		msg := new(bytes.Buffer)
		err := msg.WriteByte('?')
		if err != nil {
			return err
		}
		err = msg.WriteByte('?')
		if err != nil {
			return err
		}
		err = sendAll(msg.Bytes(), &c.conn)
		if err != nil {
			return err
		}
		msgType, msgValue, err := receiveQueryResponse(&c.conn)
		if err != nil {
			return err
		}
		if msgType[0] == 'P' {
			activeAgencies := binary.LittleEndian.Uint16(msgValue)
			log.Infof("There are %v agencies still processing", activeAgencies)
			log.Infof("Waiting before making the query again")
			time.Sleep(time.Duration(time.Duration(c.config.QueryWaitTime).Seconds()))
		} else if msgType[0] == 'W' {
			totalWinners := binary.LittleEndian.Uint16(msgValue)
			log.Infof("There are %v total winners", totalWinners)
			break
		}
	}
	return nil
}

func receiveQueryResponse(conn *net.Conn) ([]byte, []byte, error) {
	log.Infof("Waiting for server response to query")
	msgType := make([]byte, 1)
	_, err := io.ReadFull(*conn, msgType)
	if err != nil {
		return nil, nil, err
	}

	msg := make([]byte, 2)
	io.ReadFull(*conn, msg)
	if err != nil {
		return nil, nil, err
	}
	return msgType, msg, nil
}

func receiveServerResponse(conn *net.Conn) (int, error) {
	log.Infof("Waiting for server response")
	msgLen := make([]byte, 2)
	_, err := io.ReadFull(*conn, msgLen)
	if err != nil {
		return -1, nil
	}
	length := binary.LittleEndian.Uint16(msgLen)

	msg := make([]byte, length)
	io.ReadFull(*conn, msg)
	if err != nil {
		return -1, nil
	}
	log.Infof("Received response from server. %v bytes", msgLen)
	winners := processMessage(msg)
	return winners, nil
}

func processMessage(msg []byte) int {
	winners := 0
	bytesRead := 0
	for bytesRead < len(msg) {
		firstNameLenght := int(msg[bytesRead])
		bytesRead += 1
		firstName := msg[bytesRead : bytesRead+firstNameLenght]
		bytesRead += firstNameLenght

		lastNameLenght := int(msg[bytesRead])
		bytesRead += 1
		lastName := msg[bytesRead : bytesRead+lastNameLenght]
		bytesRead += lastNameLenght

		docLenght := int(msg[bytesRead])
		bytesRead += 1
		document := msg[bytesRead : bytesRead+docLenght]
		bytesRead += docLenght

		birthdateLength := int(msg[bytesRead])
		bytesRead += 1
		birthdate := msg[bytesRead : bytesRead+birthdateLength]
		bytesRead += birthdateLength
		winners += 1
		log.Infof("Winner: %v, %v, %v, %v", string(firstName), string(lastName), string(document), string(birthdate))
	}
	return winners
}

func sendContestantsInfo(contestantsList []Person, conn *net.Conn) error {
	buf := new(bytes.Buffer)
	for _, contestant := range contestantsList {
		err := addToBuffer(buf, contestant.FirstName)
		if err != nil {
			return err
		}
		err = addToBuffer(buf, contestant.LastName)
		if err != nil {
			return err
		}
		err = addToBuffer(buf, contestant.Document)
		if err != nil {
			return err
		}
		err = addToBuffer(buf, contestant.Birthdate)
		if err != nil {
			return err
		}
	}

	log.Infof("Sending batch to server")

	msgLen := new(bytes.Buffer)
	err := binary.Write(msgLen, binary.LittleEndian, uint16(len(buf.Bytes()))) //Send 2 bytes with the total length of the msg
	if err != nil {
		return err
	}
	err = sendAll(msgLen.Bytes(), conn)
	if err != nil {
		return err
	}
	err = sendAll(buf.Bytes(), conn)
	if err != nil {
		return err
	}
	log.Infof("Sent batch to server. %v bytes", msgLen)
	return nil
}

func sendAll(data []byte, conn *net.Conn) error {
	bytesWritten := 0
	for bytesWritten < len(data) {
		n, err := (*conn).Write(data[bytesWritten:])
		if err != nil {
			return err
		}
		bytesWritten += n
	}
	return nil
}

func addToBuffer(buf *bytes.Buffer, data string) error {
	err := binary.Write(buf, binary.LittleEndian, uint8(len(data)))
	if err != nil {
		return err
	}
	n, _ := buf.WriteString(data)
	if n < len(data) {
		return err
	}
	return nil
}

File: server/common/comms.py
import os
import logging
from common.utils import Contestant

def recvAll(clientSock, n):
    bytesRead = 0
    msg = b''
    while bytesRead < n:
        received = clientSock.recv(n - bytesRead)
        if len(received) == 0:
            raise Exception("Received 0 bytes from server") 
        msg += received
        bytesRead += len(received)

    if msg == b'\n\n':
        logging.info("PID {} Received finish message from {}".format(os.getpid(), clientSock.getpeername()))
        return None

    return msg


def receiveContestantsBatch(clientSock):
    logging.info("PID {} Waiting for batch from {}".format(os.getpid(), clientSock.getpeername()))
    msg = recvAll(clientSock, 2) #Receive 2 bytes with the length of the message
    if msg == None:
        return None
    msgLen = int.from_bytes(msg, "little")
    data = recvAll(clientSock, msgLen)
    logging.info("PID {} Batch Received From client. {} bytes".format(os.getpid(), msgLen))
    
    bytesRead = 0
    contestants = []
    while bytesRead < msgLen:
        firstName, bytesRead = readFieldInfo(data, bytesRead)
        lastName, bytesRead = readFieldInfo(data, bytesRead)
        document, bytesRead = readFieldInfo(data, bytesRead)
        birthdate, bytesRead = readFieldInfo(data, bytesRead)
        contestants.append(Contestant(firstName, lastName, document, birthdate))

    return contestants

def readFieldInfo(data, bytesRead):
    fieldLen = data[bytesRead]
    bytesRead += 1
    fieldData = data[bytesRead:bytesRead + fieldLen].decode("utf-8")
    bytesRead += fieldLen
    return fieldData, bytesRead

def sendWinnersToClient(winners, clientSock):
    logging.info("PID {} Sending winners to client {}".format(os.getpid(), clientSock.getpeername()))
    msg = bytearray()
    for winner in winners:
        msg.append(len(winner.first_name.encode('utf-8')))
        msg.extend(winner.first_name.encode('utf-8'))
        msg.append(len(winner.last_name.encode('utf-8')))
        msg.extend(winner.last_name.encode('utf-8'))
        msg.append(len(winner.document))
        msg.extend(winner.document.encode('utf-8'))
        birthdate = winner.birthdate.strftime('%Y-%m-%d')
        msg.append(len(birthdate))
        msg.extend(birthdate.encode('utf-8'))

    clientSock.sendall(len(msg).to_bytes(2, 'little'))
    clientSock.sendall(msg)  
    logging.info("PID {} Sent winners to client. {} bytes".format(os.getpid(), len(msg)))

